extensions/slack/src/search-messages.tsx

```diff
@@ -0,0 +1,125 @@
+import { Action, ActionPanel, Color, Icon, List } from "@raycast/api";
+import { useCachedPromise, useCachedState } from "@raycast/utils";
+import { format } from "date-fns";
+import { useState } from "react";
+import * as emoji from "node-emoji";
+
+import { withSlackClient } from "./shared/withSlackClient";
+import { getSlackWebClient } from "./shared/client/WebClient";
+import { convertTimestampToDate, handleError } from "./shared/utils";
+import { useChannels, useUsers } from "./shared/client";
+import { SearchMessagesArguments } from "@slack/web-api";
+
+function Search() {
+  const [query, setQuery] = useState("");
+  const [selectedChannel, setSelectedChannel] = useState("");
+  const [sortType, setSortType] = useCachedState<SearchMessagesArguments["sort"]>("search-sort-type", "timestamp");
+
+  const { data: users } = useUsers();
+  const { data: channels } = useChannels();
+
+  const { data, isLoading } = useCachedPromise(
+    async (query, channel, sort) => {
+      const webClient = getSlackWebClient();
+      const results = await webClient.search.messages({
+        query: `${query}${selectedChannel ? ` in:${channel}` : ""}`,
+        sort,
+      });
+      return results.messages?.matches ?? [];
+    },
+    [query, selectedChannel, sortType],
+    {
+      onError(error) {
+        handleError(error);
+      },
+      execute: query.length > 0,
+      keepPreviousData: true,
+    },
+  );
+
+  const sortOptions: { title: string; sort: SearchMessagesArguments["sort"] }[] = [
+    { title: "Newest", sort: "timestamp" },
+    { title: "Most Relevant", sort: "score" },
+  ];
+
+  const FromMeOnlyAction = () => (
+    <Action
+      title="From Me Only"
+      icon={Icon.Person}
+      shortcut={{ modifiers: ["cmd", "shift"], key: "i" }}
+      onAction={() => setQuery(query ? `from:me ${query}` : "from:me ")}
+    />
+  );
+
+  return (
+    <List
+      isLoading={isLoading}
+      searchText={query}
+      onSearchTextChange={setQuery}
+      actions={
+        <ActionPanel>
+          <FromMeOnlyAction />
+        </ActionPanel>
+      }
+      throttle
+      isShowingDetail
+      searchBarAccessory={
+        <List.Dropdown tooltip="Search Channels" onChange={setSelectedChannel} storeValue>
+          <List.Dropdown.Item value="" title="All Channels" />
+
+          <List.Dropdown.Section>
+            {channels?.flat().map((c) => {
+              return <List.Dropdown.Item key={c.id} icon={c.icon} value={c.name} title={c.name} />;
```

Comment:
issue (blocking): I am not able to look up `from:me` queries when I select Users conversations from the dropdown. For channels it works fine. Are you experiencing something similar?

Reply:
Same thing here. I've passed all the correct parameters and I'm not sure what's wrong here to be honest…

Reply:
Figured it out, the fix is:

```typescript
<List.Dropdown.Section>
  {channels.flat().map((c) => <List.Dropdown.Item key={c.id} icon={c.icon} value={'username' in c ? c.username : c.name} title={c.name} />)}
</List.Dropdown.Section>
```

Reply:
Hey @pernielsentikaer, the rest of these changes looks good to me and this is the only open issue. I have the above fix on my side but don't have permissions to push to this branch. I'm not sure how to proceed here and any guidance would be appreciated (or should we wait for Thomas to come back to this PR). This issue limits functionality in the new command, so this fix might be blocking.

Reply:
After testing it out I still see some empty lines in the dropdown, do you have a clue what might cause that?

![slack 2024-06-25 at 08 39 02](https://github.com/raycast/extensions/assets/38033107/1b6a8cab-93b8-48ca-aee6-c0dfba738d93)

It shows my name when I click one of them empty lines

![slack 2024-06-25 at 08 39 09](https://github.com/raycast/extensions/assets/38033107/5e6bc86b-daa3-4824-aefe-ee458db8d62f)

Reply:
Hey @pernielsentikaer, for direct IMs with your own username, did you send a message to yourself in slack? If so, you will need to press enter to query "From Me Only" or type the query `from:me`. You can try that query to see results from "All Channels" as well.

Apart from the direct user conversations fix, we will also need to fix the `mpim` (multi-people IM groups) search query. For that the query goes `{query} in:{mpim-name}`. The query can be `from:me` and `mpim-name` is actually the `name` (that starts with `mpdm-`) getting replaced here and we discard it:

https://github.com/raycast/extensions/blob/40c3b7048ce7bfebbeffcadb1d5a276ff1b865c5/extensions/slack/src/shared/client/SlackClient.ts#L149

You will also see this name in the detail metadata "Channel Name" for the search result. We will need to persist that value to successfully query mpim(s).

Reply:
Thanks for the fix @jfkisafk. What do you mean with the `mpim-name`? I tried using these (e.g. `mpdm-john--thomas.lombart-1`) in the query but they don't seem to work, same thing for the display name and the IDs. Any ideas what should we use here?

Reply:
Hey @thomaslombart , that is a strange behavior. Here are some things we can try:

Do you have messages from yourself in groups >= 3 folks (you + 2 others)? We can log the messages while selecting _All Channels_ and querying `from:me`. We pick the `m.channel.name` (starting with `mpdm-`) for this group of >= 3 folks and try the query `from:me in:mpdm-...`, while still keeping _All Channels_ selected in dropdown. Let me know if you don't see results.

I am able to see messages with groups of 3 if I use above steps. The problem with `mpdm-john--thomas.lombart-1` might be that it's just 2 people and when you try to start a group with one other person in slack, it switches to DM (if groups of 3 works, we might need to also hide 2 people groups from dropdown).

Reply:
Thanks for the details! It's working properly now.